k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+	// and begins resource creation the state is transformed to "starting". The
+	// object store controller observes the resources and once all resources are
+	// up and ready the state transitions to "running".
+	// The state remains in "running" until the object store is deleted, at
+	// which point the controller will clean up the associated resources.
+	//
+	//  │ object store created
+	//  │
+	//  └──►┌───────────┐
+	//      │ unknown   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller creates resources
+	//  │
+	//  └──►┌───────────┐
+	//      │ starting  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller detects all resources ready
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
+	//
+	// +optional
+	// +default="Unknown"
+	State ObjectStoreState `json:"state"`
+
+	// A list of addresses where the S3 store is exposed.
+	//
+	// +optional
+	Endpoints []string `json:"endpoints"`
+}
+
+type ObjectStoreCredentials struct {
+	// An access key. This will be the initial access key for the administrative
+	// account.
+	//
+	// +optional
+	AccessKey string `json:"accessKey"`
+
+	// A secret key. This is the secret key corresponding to the access key above.
+	//
+	// +optional
+	SecretKey string `json:"secretKey"`
+}
+
+type ObjectStoreStorageSpec struct {
+	// The initial size of the volume to provision. The volume will be created
+	// such that it can be expanded if resources are available
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:example:=100Gi
+	// +optional
+	Size resource.Quantity `json:"size"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=3
+	// +optional
+	NumberOfReplicas int `json:"numberOfReplicas"`
+
+	// Replica soft anti affinity of the volume. Set enabled to allow replicas to
+	// be scheduled on the same node
+	//
+	// +optional
+	ReplicaSoftAntiAffinity ReplicaSoftAntiAffinity `json:"replicaSoftAntiAffinity"`
+
+	// Replica zone soft anti affinity of the volume. Set enabled to allow
+	// replicas to be scheduled in the same zone
+	//
+	// +optional
+	ReplicaZoneSoftAntiAffinity ReplicaZoneSoftAntiAffinity `json:"replicaZoneSoftAntiAffinity"`
+
+	// +optional
+	DiskSelector []string `json:"diskSelector"`
+	// +optional
+	NodeSelector []string `json:"nodeSelector"`
+
+	// +optional
+	FromBackup string `json:"fromBackup"`
+
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default:=2880
+	// +optional
+	StaleReplicaTimeout int `json:"staleReplicaTimeout"`
+}
+
+type ObjectEndpointSpec struct {
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// +optional
+	DomainName string `json:"domainName"`
+}
+
+type ObjectStoreSpec struct {
```

Comment:
As mentioned above, I am proposing to have a new field in the spec: `spec.DesireState` so that user can voluntarily stop an ObjectStore